### ISSUE TITLE
New `tap` subcommand v1

### ISSUE THREAD

**Issue by binarylogic** [timberio/vector#3212]
As part of https://github.com/timberio/vector-product/issues/24, we want to introduce a way for users to tap into a running Vector instance and sample events.

## Goals

1. Ability to sample events at any step within the configured pipeline.
2. Do so in a way that will unblock the web UI. The web UI will have sampling ability.

## Proposal

I propose that we introduce a new `vector tap` subcommand:

```bash
$ vector help tap

USAGE:
    vector tap [OPTIONS] [ID]

OPTIONS
    --limit            Exit when this limit is reached (no default)
    --position         Whether to tap at the front or back of the component (default front)           
    --sample-rate      How often events should be sampled (default 1s)

ARGS
    ID                 The component ID to tap.
```

And usage is simple:

```bash
$ vector tap my-files-source
0" 406 26004
86.143.250.111 - - [26/Jul/2020:13:09:43 -0400] "POST /engineer HTTP/1.1" 502 5575
164.149.249.195 - - [26/Jul/2020:13:09:43 -0400] "POST /extend/incubate/roi HTTP/2.0" 100 11738
57.57.164.242 - tillman5416 [26/Jul/2020:13:09:43 -0400] "HEAD /architectures/e-tailers/clicks-and-mortar HTTP/1.1" 416 14004
234.49.218.76 - - [26/Jul/2020:13:09:43 -0400] "GET /relationships HTTP/1.0" 504 19698
61.188.31.127 - - [26/Jul/2020:13:09:43 -0400] "HEAD /harness/cultivate/seamless/b2c HTTP/2.0" 205 18781
22.61.249.23 - feest7241 [26/Jul/2020:13:09:43 -0400] "POST /partnerships/grow/impactful/deploy HTTP/2.0" 302 13686
139.114.84.146 - schulist2106 [26/Jul/2020:13:09:43 -0400] "PUT /target/web-enabled/b2b HTTP/1.1" 403 26476
```

And tapping the output:

```bash
$ vector tap --position=back my-files-source
{"timestamp": "...", "message": "86.143.250.111 - - [26/Jul/2020:13:09:43 -0400] "POST /engineer HTTP/1.1" 502 5575", "file": "..."}
```

**Comment by MOZGIII** [comment]
I was looking for something like this for the K8s E2E tests. My use case was to assert that the produced output of the sink matches the expectation from an externally driven test flow. I'm looking forward to using something like this!

**Comment by binarylogic** [comment]
I'm removing this from the initial Vector CLI observability work until we can gather more requirements. There's some overlap here with some of the fleet management/distributed pipeline building tasks.

**Comment by leebenson** [comment]
To start gathering some requirements, I've laid out a few assumptions below. Nothing here is set in stone or intended to be a complete list -- this just a starting point. Feel free to add/tweak/suggest:

- The initial version targets a single Vector instance. I don't think we need to get bogged down designing for possible overlap with fleet management for this initial version. A multi-Vector version is likely to introduce one or more infrastructure elements that aren't applicable to a single Vector instance; it seems unfeasible to try to target both until we have a clear roadmap.

- Like `top`, we should be able to pass a `--url` param to sample events from a remote instance.

- Our existing API tooling facilitates a pub/sub pattern that should be ideal for real-time events, via GraphQL subscriptions. This would work in a similar way to streaming metrics.

- The UI should be able to tap into an event stream by passing a GraphQL query. This would be the same query used by `tap` to retrieve events over a WebSocket.

- We have the ability to support two modes of capture: front/back positioning (where events are captured as a 'snapshot' over a given interval, and X first/last records are returned), or streaming events, where every event -- as it occurs -- is surfaced to the client (subject to other parameters, such as `--limit`.) @binarylogic - your issue description suggests the former, but it probably wasn't very clear last July that we'd have streaming subscriptions to leverage, so does this alter/add to requirements any? FWIW, if we want to make real-time events an option, I think that's feasible without requiring a separate implementation.

- Any component that returns data in a serde serialisable format should work for returning data to the console. Should the format be user-definable -- e.g. plain text/json/yaml, etc?

- For the first version, I propose queries that return message data as a string. This provides for the UX of a tailing console, for both the CLI and the UI. If we want to return semantic meta, that could later be expanded per-component, depending on the capabilities of the component itself to derive meaning from the event data being returned. 

- Visually, I'm assuming the CLI to continue running and printing out events to stdout until `--limit` is reached or SIGINT is received, whichever is sooner. The equivalent mechanism would occur for all GraphQL clients, resulting in a completed subscription.

**Comment by jszwedko** [comment]
Thanks for putting these together!

This could also be an enhancement, but I think the ability to sample the data would be useful for high volume streams; including being able to sample data differently based on the data (for example show me all HTTP 500s, sample 200s).

> For the first version, I propose queries that return message data as a string. This provides for the UX of a tailing console, for both the CLI and the UI. If we want to return semantic meta, that could later be expanded per-component, depending on the capabilities of the component itself to derive meaning from the event data being returned.

Do you mean just the `.message` key of messages?

ccing @lukesteensen for visibility since I feel like he talked about this feature when I first started at Timber and probably has opinions :smile:

**Comment by leebenson** [comment]
> This could also be an enhancement, but I think the ability to sample the data would be useful for high volume streams; including being able to sample data differently based on the data (for example show me all HTTP 500s, sample 200s).

👍 I like this idea. We could have multiple filters per query, each with their own velocity. Perhaps something like:

```gql
subscription Events($on: EventsOnInput!) {
  events(on: $on) {
    ...
  }
}
```

With:

```json
{
  "on": [
    {
      "name": "http_errors",
      "filter": {
        "statusCode": {
          "equals": 500
        }
      }
    },
    {
      "name": "http_hits",
      "interval": 2000,
      "filter": {
        "statusCode": {
          "equals": 200
        }
      },
      "first": 10,
      "limit": 1000
    }
  ]
}
```

In this case, the first filter would return all `500` status codes live from the `http_errors` component, for the duration of the subscription.

The second would return the first 1,000 `200`s from `http_hits`, capturing at 2s intervals, returning the first 10 of each batch.

Is that the kind of thing you had in mind?

> Do you mean just the .message key of messages?

I was actually referring to the _full_ message, rather than a specific `message` key on it. The idea being, where we have a component emitting event data that is serializable, in this first instance we'd simply pipe it to stdout per the default/provided format.

I haven't given this much thought as to its feasibility; it may well be that we only want certain event types matching a known shape to be a candidate for `tap`. Perhaps that's required anyway, if we're going to have filters that can discern, for example, that a status code is 200 vs. 500. In which case, it may make sense to return semantic types via the API rather than just a dump of the raw event.

**Comment by lucperkins** [comment]
One possibility for defining sample sets would be passing in VRL conditions as a string. That would be a way of avoiding `check_fields`-style logic.

**Comment by leebenson** [comment]
I like the idea of using VRL in more places. I think there's a couple of ways we could go with this:

1. Build a translation layer that converts VRL -> GraphQL filters. This would have the benefit of parity with the typed filters that we're using as part of the filtering/pagination/sorting interface.

2. Take a free-form VRL string that's used in lieu of GraphQL filters. These could be applied more generally against the result set, before being passed to the Relay pagination filters (where appropriate.)

In either case, I'd imagine input would be a GraphQL scalar, with ad-hoc compilation of the VRL query occurring at runtime, and errors being passed back in the scalar validation phase. This would keep the type as a string for calling clients, and ensure we're working with a valid query before hitting the relevant field resolver.

I'm not sure which is the more pragmatic approach. @JeanMertz, @FungusHumungus -- any thoughts?

**Comment by FungusHumungus** [comment]
@leebenson I'm not very familiar with GraphQL. Can you expound a bit on what a GraphQL filter looks like and how it could be expressed in Rust?

**Comment by leebenson** [comment]
Sure @FungusHumungus  👍 

GraphQL is a type-safe query language.

Beyond basic scalar values, all types -- for input and output -- are defined by us. There are no built-in concepts for filtering, sorting, etc. But whatever types we create, are known at compile-time and are exposed to connecting GraphQL clients so that both the client and server can validate the query before attempting to process it.

Because of this type rigidity, we wind up with filters being generated during compilation that wind up looking like this:

![image](https://user-images.githubusercontent.com/719722/105022175-57f54d00-5a41-11eb-977f-e7fa300d8648.png)

In the case of filtering components, we can then issue a query like this:

```gql
{
  components(filter:{name:{equals:"wifi"}}) {
    edges {
      node {
        __typename
        name
      }
    }
  }
}
```

Here, the `name` field is known in advance, as is the `StringFilter` type that provides us with `equals` and a set of other operations.

The base filter utils are in https://github.com/timberio/vector/blob/master/src/api/schema/filter.rs; this particular `ComponentsFilter` is then implemented in https://github.com/timberio/vector/blob/master/src/api/schema/components/mod.rs#L57-L80 and used inside the `components` query at https://github.com/timberio/vector/blob/master/src/api/schema/components/mod.rs#L94-L97.

As GraphQL is the API gateway/interface we use for `top` and for the UI's comms with Vector, my intention is to use the GraphQL for `tap`. Specifically, we'd be using GraphQL subscriptions, which enable data streaming back to the client via WebSockets (which, again, we also use for `top`.)

To use VRL with GraphQL, I therefore think we could do one of two things:

1. Keep the existing GraphQL filters, and attempt to map a VRL query to one of the types (such as `ComponentsFilter`, in this case.) I'm not sure how this would be done. Perhaps implementing a trait for ComponentsFilter that indicates the fields and allowed operations, which in turn is used as validation for GraphQL input. I can solve the latter-- what I lack the context on is how we might do that 'mapping'. Is there an API we might leverage for transforming a VRL expression into a struct?

2. Ignore our custom GraphQL filters entirely, and instead just accept string input where a VRL query is used. This could still be type-safe; the GraphQL filter would, in this case, be a union of something like `ComponentsFilter | VRLExpression`, where `VRLExpression` is evaluated at runtime and executed against the result set to return filtered items. Again, I lack the context to know exactly how this would work, but the point is that this option wouldn't require us to 'map' to the custom filter types at all... it would operate directly on the results set being returned to the client, as a kind of post-filter.

So, taking GraphQL out of the equation for the moment, I think the questions are:

-  Is the kind of filtering query @lucperkins provided as an example, generally a good match for VRL?
- Can VRL be used to filter an arbitrary iterator?
- Is there an API that allows us to limit fields/operations to a particular field set or scope? (such as `code` and `level`, in @lucperkins' example.)
- Would it be simpler to define rules that map a VRL operation to a struct of a known shape... or use it directly as a filter against an iterator?

Based on the above, I think it will be clearer how to make the VRL and GraphQL pieces interact.

**Comment by FungusHumungus** [comment]
Beautiful, thanks.

There are two stages to Remap.

# Compile.  

First the program is compiled by calling [Program::new](https://github.com/timberio/vector/blob/master/lib/remap-lang/src/program.rs#L102-L107), see [the Remap conditions](https://github.com/timberio/vector/blob/master/src/conditions/remap.rs#L43-L48) for an example. 

The type constraint (eg [here](https://github.com/timberio/vector/blob/master/src/conditions/remap.rs#L24-L31)) specifies what type of result we are expecting from the program. For conditions, they have to resolve to a Boolean.

Note we do as much as possible in the compile stage and we aren't optimising for performance here as it is designed to be done once at boot time.

# Run

With a successfully compiled program, we can run it (eg [here](https://github.com/timberio/vector/blob/master/src/conditions/remap.rs#L75-L78)). The program must be passed an object that implements the [Object trait](https://github.com/timberio/vector/blob/master/lib/remap-lang/src/object.rs#L5). That is the state the program can operate on. 

Currently the script can mutate that object. We plan to create an immutable mode at some point, but until then the Condition creates a clone of the object to prevent the condition from producing any side effects.

----------

So given that, I would suggest VRL would be a good fit for GraphQL if you can:

1. Compile the filters up front. Potentially each tap query could start by compiling any remap and then used the compiled programs in the loops. How often are the queries going to be run? Can they afford a performance hit at the start of each query? The performance hit I'm talking about would be around the same order of magnitude or slightly more as compiling a regular expression.
2. Implement `Object` for the data you are running the filters against.

**Comment by leebenson** [comment]
Great, thanks for the overview. Super helpful.

Based on this, it looks like it ought to be feasible to translate a remap query to a GraphQL filter struct. If that's the case, we wouldn't need any changes in the API. I'll have a play with it.

I'm not concerned with runtime compilation perf. GraphQL input is validated once at runtime, when a query is sent.

I'm imagining we'll wind up with a `VRLFilterExpression` (or similar) type that compiles into a filter. It might add a couple of hundred ns to the initial query response, but there's no further penalty for subsequent results back to the client.

**Comment by leebenson** [comment]
Here are the points I could do with consensus on, before getting too far:

### - Is `tap` intended to be applied to _all_ components? Or a subset of them?

Considerations:

- Are all components guaranteed to have a serializable representation of event data, fit for stdout?
- Is there a way to generalise event data so it can be filtered/printed to the console?

### - Along with capturing events at a user-definable interval (and selecting first/last x), do we also want to offer live tailing

FWIW, I don't expect live streaming of events to add any additional complexity beyond capturing events over an X ms interval, and emitting Y from the front/back position. There are existing equivalents for this in the API with metrics that use the same resolvers and helper methods-- I think that would be the case here.

The question is more of a UX/user requirement, vs. implementation cost.

### - Do we want to allow for multiple filters?

e.g. per @jszwedko's earlier example -- live `500` status codes as they happen, `200` codes sampled at X ms interval.

In this specific case, each filter would require its own interval, search query, and limit; the former would be streamed in real-time, the latter might be sampled every couple of seconds, and only tail the latest result.

Or should one `tap` session = one filter?

### **- Like `top`, should `--url` observe remote events?**

Again, there are existing patterns for this, so the complexity cost is minimal. Down to our/user expectations.

### **- Should filters be definable in VRL as well as structured types?**

See some of the discussion above. This is the biggest unknown right now, but one that - on the surface - seems doable.

If we want this, do we want this from v1... or as an upgrade path?

**Comment by leebenson** [comment]
Of the aforementioned, I think the one I could do with the earliest direction on is whether we're targeting all components -- or a specific subset.

I think this is likely to inform the strategy for getting events out of topology, and into the API for consumption.

I've been hunting for a good integration point. There are a couple of options so far, assuming we want to observe _everything_ that's happening across a pipeline:

### 1. Treat `tap` as an internal sink

This is similar to the path I initially started down with the API, before switching to it being an external layer that simply reads topology config. The benefit to it being a sink is that it could integrate with the regular fan-out pattern without changing topology.

The downside is essentially that, outside of the existing broadcast mechanism, it's not really a sink.

We'd be mashing together the concept of an internal receiver channel, with something that's user-configurable... and has several characteristics that we don't really need to design for-- such as ack'ing event reception, or requiring a string name that could clash with what's being defined in config.

It'd also require explicitly defined inputs, which don't really map in the case of being able to tap _all_ components, since we'd actually also want to tap the sinks themselves.

For those reasons, I'm not sure this feels appropriate.

## 2. Modify topology to broadcast events to a channel

When any component produces an `Event`, it would essentially be copied down a channel. This is probably the easiest, since we'd receive a copy of whatever's flowing through the topology pipeline. The listening end would be an API subscription -- e.g. a `tap` user, or any other client using the GraphQL API with an active subscription in-flight.

The downside to this, is that it could be expensive. We're sending data down a channel that may not have an active listener 99%+ of the time.

There are perhaps a couple of ways we could mitigate this (these are design options, not performance recommendations-- I have no idea how well they'd perform without testing):

1. Have an `Arc<Option<Receiver<Event>>>`, which pipes data to the channel only if there's a listener. When at least one subscription/`tap` user is active, an `Event` is sent down the channel to be consumed/filtered by the relevant API resolver.
2. Have some way at the call site of broadcasting events to know what is being requested, and only send events that match. This is potentially quite tricky, since it's placing the onus on the sender rather than the receiver, which itself would need to know what conditions to send an event on. These conditions aren't likely to be known until the query is executed.
3. Maybe having an open channel isn't actually that expensive if we don't have an active listener. Send the data along, running with the assumption that the majority cost is on reading/filtering the data, which is done in the relevant API resolver. This would be the simplest to integrate.

My main concern is that `tap` is likely to be a minority % of the total usage of a Vector agent, so paying the cost of dispatching event data to a receiver that's used infrequently, and even then only typically against select components, may result in a performance hit the majority of time Vector is running without being observed.

## 3. Do the same as 2, but outside of topology

There may be some other route in that I'm not yet considering that exists outside of topology, in the same way that the API is itself external, and reads config and metric data directly.

I'm still exploring the code base for a relevant integration point.

Turning the question to the team: with the goal of observing events produced by all components before (or after) they flow to transforms and sinks... is there any obvious places you'd insert a listener to echo those _out_ of topology? This should include sinks themselves. cc @lukesteensen @bruceg

**Comment by leebenson** [comment]
In the interim, I'm currently experimenting with piping events through to an API channel in places like this https://github.com/timberio/vector/blob/5efa5d95f48640295728ac4c0672cfe6545b43ed/src/topology/builder.rs#L126

**Comment by jszwedko** [comment]
Thanks for the write-ups @leebenson !

Some thoughts:

I think we could start with just sampling events flowing through the topology rather than data flowing into sources and out of sinks as that is likely to be more complicated, it's unclear to me how we'd apply filters to them, and the serialization format is ambiguous. This means we'd instead allow you to tap in one of four places: after a source, before a transform, after a transform, before a sink. One advantage is that this should let you keep the logic isolated in the topology.

That should also answer the question of "Are all components guaranteed to have a serializable representation of event data, fit for stdout?" in that you'll just be dealing with `Event`s and you can serialize them however you like (or maybe let the user decide): JSON, logfmt, etc..

Just a note that sampling on an interval by taking the first X or last X might cause the users to see an unrepresentative sample since it never samples events outside of that range during the interval. I'm imagining an extreme case where some event happens in their app every 1 second that outputs a bunch of logs. If you sample the first X events every 1 second you will only ever see those logs and miss the others. I think we could probably just implement the sampling similar to the `sampler` transform in that all events pass through and are sampled at a given rate. You might even be able to share code with the `sampler` transform. One feature it has, that could be useful in the `tap` sampling context, is the `key_field` that can be used, for example, to sample in a way that you get all of the logs for a given request_id.

:+1: to `--url` support here too for remote tapping.

I'll let others comment on the overhead of always publishing to channels with no listeners, but I'd probably just do it in such a way that the listeners can be added on-demand to the topology.

I think live tailing is useful for low volume streams.

**Comment by leebenson** [comment]
Thanks @jszwedko, appreciate you helping me dive into this!

> I think we could start with just sampling events flowing through the topology rather than data flowing into sources and out of sinks as that is likely to be more complicated, it's unclear to me how we'd apply filters to them, and the serialization format is ambiguous. This means we'd instead allow you to tap in one of four places: after a source, before a transform, after a transform, before a sink. One advantage is that this should let you keep the logic isolated in the topology.

👍 Great, that sounds very similar to where I landed so far. As I was exploring some of the questions above, it wasn't immediately clear whether there could be component-specific expectations around observing data, or whether we're just talking about exposing a `Event::Log` to a subscriber.

Assuming the latter, it seems that doing a passthrough `.inspect` a la https://github.com/timberio/vector/blob/5efa5d95f48640295728ac4c0672cfe6545b43ed/src/topology/builder.rs#L126 would work quite well to bubble that up to the API without having to hack at topology itself.

> That should also answer the question of "Are all components guaranteed to have a serializable representation of event data, fit for stdout?" in that you'll just be dealing with Events and you can serialize them however you like (or maybe let the user decide): JSON, logfmt, etc..

Right, that makes sense given an `Event`. As above, I wasn't sure whether we were thinking we'd be working with raw bytes flowing through components, or structured logs.

> Just a note that sampling on an interval by taking the first X or last X might cause the users to see an unrepresentative sample since it never samples events outside of that range during the interval. I'm imagining an extreme case where some event happens in their app every 1 second that outputs a bunch of logs. If you sample the first X events every 1 second you will only ever see those logs and miss the others. I think we could probably just implement the sampling similar to the sampler transform in that all events pass through and are sampled at a given rate. You might even be able to share code with the sampler transform. One feature it has, that could be useful in the tap sampling context, is the key_field that can be used, for example, to sample in a way that you get all of the logs for a given request_id.

That's a good point. We probably want to sample a more even distribution, by default. @binarylogic, how does this fit with your thoughts on defining an explicit front/back position?

**Comment by lukesteensen** [comment]
@leebenson 

Thanks for writing this up! It seems like you're hitting on all the right questions, so I'll try to get my own thoughts down here. Keep in mind they're relatively high level and I haven't tried to implement anything yet, so take them with an appropriately sized grain of salt :smiley:  

> * Are all components guaranteed to have a serializable representation of event data, fit for stdout?
> * Is there a way to generalise event data so it can be filtered/printed to the console?

Yes, I think of this as pretty much the same as the ability to connect a `console` sink at any point in the topology. I don't think we need to or should really try to deviate from that at all. As much as possible, let's try to make it just fit in naturally with the way we do everything else.

> Along with capturing events at a user-definable interval (and selecting first/last x), do we also want to offer live tailing?

It's not entirely clear to me what "capturing events at a user-definable interval" means, and it does seem more complicated than "just" live tailing. Again, I think the simplest mental model for this feature is the same topology-level input connection you'd get if you configured a sink in your TOML file. We should certainly support flexible filtering, sampling, etc, but that can be mostly an implementation detail of the "sink" and not affect the way that it interfaces with the rest of the system.

> Do we want to allow for multiple filters?

As much as possible, I think this should align with what you could otherwise configure manually via TOML. It might be tricky to support right off the bat, but it'd be great to let users supply any combination of transform that they want (e.g. remap, sample, reduce, etc).

> The downside is essentially that, outside of the existing broadcast mechanism, it's not really a sink.
> 
> We'd be mashing together the concept of an internal receiver channel, with something that's user-configurable... and has several characteristics that we don't really need to design for-- such as ack'ing event reception, or requiring a string name that could clash with what's being defined in config.

I'm not sure I agree with this. It is a sink in that it "consumes" events. That's pretty much all there is to being a sink. They are internal receiver channels and user-configurable. Acking won't be any different than the console sink and naming shouldn't be a problem with a little random number generation. In my opinion, making something that's not sink-shaped would require passing a pretty high bar and it doesn't seem like we're there.

> It'd also require explicitly defined inputs, which don't really map in the case of being able to tap _all_ components, since we'd actually also want to tap the sinks themselves.

I'm quite sure that we will want to require that users tell us what to tap (i.e. specify inputs). I don't think a default of all components would be very intelligible or useful. As for sinks, I don't see a feasible way to tap their "output" or internal state, since it varies dramatically and would require significant work to invent something to output for each sink. The simpler equivalent in my mind would be to tap the same set of inputs as the sink, to see the data that it is ingesting.

> Modify topology to broadcast events to a channel

I am pretty strongly against any solution that significantly impacts the behavior of our topology when this feature is not in use. Always fanning out every event would certainly qualify.

Luckily our topology already has the ability to dynamically start fanning out events to a new destination via config reloading. My preferred strategy would be to build on that existing functionality, generalize it a bit, and try not to introduce entirely new mechanisms into what's already a pretty complicated part of the system. Another benefit there is that it pretty naturally supports adding whatever kind of pipeline you want in front of the actual output to the user, so we could avoid reinventing things like sampling and reuse whatever we do build as normal transforms.

**Comment by leebenson** [comment]
Thanks for all of the above @lukesteensen.

I think the main issue with treating `tap` as a sink, is that its purpose is intended to be ad-hoc; activated by an incoming query, and not a permanent fixture in vector.toml.

Two uses cases are:

1. Sampling `Event`s via a CLI subcommand. A typically usage might be something like `vector tap <component> --url <remote> --interval 2000 --filter ".status == 500" --limit 5`. Unpacking this a little, this means remote observability (ideally, via the GraphQL API over WebSockets, to also work in the UI), a filter that applies only to this one active request (other users could have different queries in-flight), and a sample over a 2s interval. Other in-flight queries could have different filters and velocities -- all the way to real-time.

2. Real-time views in the UI. Multiple users clicking around the web interface, zooming in to different pieces of topology to sample the `Event`s they're sending, getting back structured data that can be used to populate tables, etc.

Both usage cases fire sporadic requests to the GraphQL API that dip in and out of this data as required.

Because of this dynamism, I'm not sure this being statically configurable via vector.toml answers the above use-cases.

We could certainly use config to limit the scope of the components we wish to make available for observation, or perhaps whitelist certain filters or other tweaks, but I think the purpose there would be more for security/setting scope than it would directly address the mechanics of the feature.

After some more thought (and thanks to some input from @jszwedko above), I _think_ the intention for this feature is to limit observation to `Event`s (typically `Event::Log`-- other parts of the GraphQL schema already address metrics). If that's the case, I think we can actually treat this as just another part of the API and not touch _RunningTopology_ at all. In the same way that the API captures metrics ad-hoc, I think it can capture events ad-hoc.

To make the implementation I have in mind a little clearer, I'm hoping to put together a quick demo today that walks through an example use-case and where I think that integration point might make sense. That should hopefully clarify the scope a little and tease out the requirements.

**Comment by leebenson** [comment]
@binarylogic -- solely from a UX perspective: what's your expectation for enabling this feature?

There are effectively two options, which I think (other factors notwithstanding) will help determine whether or not this ought to be implemented as a sink:

## 1. Explicit `inputs` via vector.toml

This option requires a user to select which of the components they wish to observe ahead of time, in config.

A typical config might look like this:

```toml
[api]
  enabled = true

[sources.gen1]
  type = "generator"
  lines = ["{\"count\": 1}"]
  batch_interval = 0.01
  format = "shuffle"

[sources.wifi]
  type = "file"
  data_dir = "checkpoints"
  include = ["/var/log/wifi*.log"]

[transforms.json]
  type = "json_parser"
  inputs = ["gen1"]

[sinks.blackhole2]
  type = "blackhole"
  inputs = ["json", "wifi"]
  print_amount = 1000

[sinks.tap]
  type = "tap"
  inputs = ["generator", "wifi", "json"]
```

**Notes:**

- The 'tap' sink explicitly defines the components that will emit `Event`s to it.
- A query to the GraphQL API filters those events by the criteria the user specified (specific components, VRL filter applied, etc), at the interval the user specifies.
- There's no way to observe other sinks, since this is part of the regular _source -> (transform ?-> transform) -> sink_ pipeline.
- The 'tap' sink is only valid if the API is also enabled. Otherwise there's no way to get the data out.

Based on a cursory exploration so far, the pros and cons as I see them are:

**Pros:**

- It uses vector.toml that the user is already likely familiar with, and is using to configure most everything else.
- It's explicit which sources/transforms can be observed; the user has to opt-in at run time.
- It's performant. We're not shipping an `Event` down a channel that 99% of the time, may have no active subscribers.
- It inherits the work and characteristics of the topology pipeline, which has been tuned for performance, handles SIGHUP/graceful shutdowns, etc.
- After config, the GraphQL API can work in exactly the same way as the 'no config' option. Users can define capture intervals, filters, etc ad hoc, in exactly the same manner.
- Avoids hacking at existing topology, or finding some integration point to latch onto, that wasn't designed for this use case.

**Cons:**

- It requires every component to be defined as an explicit `inputs` item.
- It becomes an opt-in feature. A user with the API enabled but not the sink tap, wouldn't get any data.
- It's potentially confusing for UX to query for a component you know exists, but wasn't explicitly set in `inputs`.
- Sinks can't be observed. Per @lukesteensen's comments, that may not make any sense anyway, as the `Event` is generated from sources/transforms.

## 2. No configuration required

In this option, `api.enabled` must be on -- but additional config is skipped.

The user can observe any configured component's emitted `Event`s (where relevant), and -- like the first option -- these can be queried ad hoc.

**Pros:**

- No explicit configuration is required.
- All components are automatically visible; no opt-in necessary (this could also be a con.)
- No extra code for defining a sink.

**Cons:**

- All components are automatically visible to the API, since they weren't explicitly opted in via config.
- Performance concerns. All events from all sources would be emitted to a channel that may only infrequently have subscribers.

----

@binarylogic - which of the two options match your expectations for this feature?

**Comment by MOZGIII** [comment]
Another con for the "Explicit inputs via vector.toml" - if implemented as a regular sink, the events will all be mixed up, and we won't be able to show the metadata like what input the event had arrived from.

**Comment by leebenson** [comment]
>Another con for the "Explicit inputs via vector.toml" - if implemented as a regular sink, the events will all be mixed up, and we won't be able to show the metadata like what input the event had arrived from.

Thanks @MOZGIII, that's very useful feedback. I hadn't considered preservation of meta. That'll certainly be required, since a query will be against specific component(s).

> The "No configuration required" option could only start piping the data into its flows if an active subscription exists, and we could filter by the components we tap into at the subscription level. Alternatively, we could just tap into a specific component - exactly one at a time, at the GQL API level (after reading some more, I understand this is implied already).

Yes, this is what I had in mind. Effectively, _all_ `Event`s would be sent through a channel. When an active subscription is made, the channel (provided via request context; an async-graphql feature) would create a receiver, which would then filter those events based on the filters/params provided in the query.

This is effectively how we already do it with metrics. The difference is that metrics are captured globally, vs. creating a per-request receiver.

My one concern with this approach was/is the performance of sending data through a tokio channel, when there are no active listeners. I understand that'll return an `Err` in each case (which is normal behaviour when we don't have any listeners, and can be ignored), but given an in-flight subscription probably happens only a tiny fraction of the time, is seems potentially expensive to be attempting to send down a dead channel for every Event.

(I just saw your latest edit now addressing this... thanks!)

> Also, performance is a huge concern here. I think we should ideally use something like ringbuffers (or try_send in channels) to ensure we're never stuck - losing the events in the tap flow if they can't be flushed to the API in time seems like a preferable option to back pressuring the whole topology flow.

Yep, agreed. Losing events is perfectly acceptable. There doesn't need to be any back pressure. A `tap` operation is ephemeral by nature; we only care to capture at the time of request. Ultimately, I think the question is -- what's the cost of _attempting_ a send if there are no listeners on the other end?

**Comment by MOZGIII** [comment]
> Ultimately, I think the question is -- what's the cost of _attempting_ a send if there are no listeners on the other end?

I would view this as generally high - as attempting to send will require us to obtain a `.clone()` of the `Event`. We should try to optimize this away by avoiding the attempt in general. I.e. a branch (`if some_bool`) conditionally enabling sending the copies of the `Event` at the processing of each `Event` before/after each component interconnection at the topology is probably fine, but sending the event over a blackhole channel if there are no consumers is way too much of a price to pay for nothing.

**Comment by leebenson** [comment]
That makes sense. So some condition where we could determine whether there's an active channel without actually attempting the channel. Maybe an atomic bool when a subscription is activated?

**Comment by MOZGIII** [comment]
Btw, I agree with @lukesteensen that we should try to make use of the topology reload somehow to avoid introducing another mechanism. That is to a degree though, because maybe it would really be more straightforward. However, it is just hard to tell. in such cases, I usually consider it fine to implement both options, compare, and either pick one of them, or go with a new iteration (a third option). In this case, the choice is between relying on the topology reload, or introducing a new mechanism for runtime switching.



> That makes sense. So some condition where we could determine whether there's an active channel without actually attempting the channel. Maybe an atomic bool when a subscription is activated?

Yes, maybe something like this if we choose not to go with the topology reload. `ArcSwap<Option<Sender<...>>>` maybe? Not clear.

If we choose to go with the reload - it will probably be just `Option<Sender<...>>` - or a regular, non atomic `bool`.

**Comment by leebenson** [comment]
> Btw, I agree with @lukesteensen that we should try to make use of the topology reload somehow to avoid introducing another mechanism. That is to a degree though, because maybe a new mechanism would really be more straightforward. However, it is just hard to tell upfront. In such cases, I usually consider it fine to implement both options, compare, and either pick one of them, or go with a new iteration (a third option). In this case, the choice is between relying on the topology reload, or introducing a new mechanism for runtime switching.

I actually don't think we need to worry about the reload mechanism, if we opt to place the Event emission somewhere like [here](https://github.com/timberio/vector/blob/5efa5d95f48640295728ac4c0672cfe6545b43ed/src/topology/builder.rs#L126).

We only really need to care about Events being emitted and picking them up in the API, which will only happen if the relevant component is loaded and the topology pipeline is already initialised. Reloading/teardown is external to that.

This is how the API already works. It collects metrics globally, but for those metrics to get there, the topology pipeline has to be active. The API remains outside of it.

Unless I'm missing something, I think that would be the case here?

**Comment by MOZGIII** [comment]
Well, the way I understood it was that there are two ways we can do it:
1. We pass some API-driver parameter (something like `tapped_components: Vec<(String, Sender)>` - string being the component name here) at the topology builder level - that will then dynamically change when clients subscribe/unsubscribe, and we'll be reloading the topology with the new value every time this happens; this way we rely on the topology reloads for synchronization.
2. The topology itself, after being initialized, provides a call in the form like `fn tap(&mut self, component_name: &str, ) -> UntapOnDrop<Receiver>;`, where `UntapOnDrop` is some magic wrapper that would untap the component when the sender is dropped. Or, something more manageable - like `fn tap(&mut self, component_name: &str, sender: Option<Sender>) -> Option<Sender>;` (where `Option<Sender>` in the arg is the sender to use on `None` if it's not used, and in the return value is whatever the previous value used to be). You get the idea - some new interface to allow already built topology to start/stop tapping a particular component - for instance by updating the `ArcSwap` mentioned above.

**Comment by leebenson** [comment]
Gotcha @MOZGIII, that's interesting.

I was thinking the flow could be:

```
Event -> API Sender -> API Receiver (subscription) -> Filter based on params -> Out to the client
```

But what I think you're suggesting is actually having a call to a `RunningTopology.tap()` that receives the Sender ad hoc, transmits events back for the duration of an active channel, and closes afterwards?

I like that approach. That scopes cloning the event to an active request, which seems like it'd be a lot more efficient.

FWIW, the API actually returns a `Stream<Item = T>` back to the client for WebSockets. I'm wondering if we could exploit that to avoid channels altogether, by returning a stream of Events matching the component name in `.tap()`, and then further filtering in the subscription resolver.

Once the subscription terminates (explicitly on the client, or the connection dies), the stream would cancel, effectively freeing the ad hoc listener.

How does that sound?

**Comment by jszwedko** [comment]
I'll let @lukesteensen confirm but I don't think he was suggesting literally configuring tap as a `sink` in the TOML file. It's just that tap behaves like a sink so it could be modeled as a (virtual) sink in the topology that is never persisted to the config file; it is just injected into the running topology when `vector tap` is used and removed when it disconnects. An advantage of this, is that you could string together arbitrary transforms to marshal the data how you want to tap it (future work).

**Comment by leebenson** [comment]
I'll let @lukesteensen confirm, but this line seemed to suggest an explicit config:

> I'm quite sure that we will want to require that users tell us what to tap (i.e. specify inputs)

**Comment by jszwedko** [comment]
:sweat_smile: I feel like I read that differently. I believe that is just saying we should have the user tell us exactly where they want to insert the "tap" rather than tapping all inputs.

> I'm quite sure that we will want to require that users tell us what to tap (i.e. specify inputs). I don't think a default of all components would be very intelligible or useful.

**Comment by leebenson** [comment]
Either way, I definitely agree taps should be selective. Seems wasteful to send everything when there's no-one listening!

**Comment by binarylogic** [comment]
Apologies if I missed anything above, there’s a lot here. I’m slightly concerned that either a) the feature is much more complex than anticipated b) we’ve made it more complicated than it needs to be. Either way, the value of this feature is not worth the complexity that I’m sensing, which brings me to the use case. The use case is production setup and debugging. The ability to inspect events running through a Vector instance to investigate abnormal behavior, usually due to malformed data. That’s it really. This feature is _very_ important for a GUI intended to build production pipelines, but it sounds like we may solve that differently?

Given the above, a light RFC is in order. If we find that the RFC is overly complex, or the discussion is heavy, we should consider not doing this project as it’s value would not offset the cost.

**Comment by jszwedko** [comment]
:+1: I agree that an RFC would be useful for this feature.

In addition to the use case you mentioned of debugging vector topologies, I'd like to add another use case which is the ability to live tail logs during events like a production deploy or during an outage a log service provider.

**Comment by lukesteensen** [comment]
Sorry, I definitely did not mean that you'd specify a literal `tap` sink in your TOML config file. My point was that _conceptually_ it doesn't need to be any different. We already have dynamism via config reloading, and I think we'd be better served by solidifying that existing functionality than adding a new mechanism.

You can imagine the dumbest possible version of this being an API handler that accepts a TOML snippet, appends it to the config file, and triggers a reload. Then when the connection is closed or times out, it removes the addition and triggers another reload. That's obviously not how we'd actually implement it (and a bit of a simplification), but it hopefully helps illustrate that there's not necessarily anything fundamentally different going on.

**Comment by leebenson** [comment]
> Apologies if I missed anything above, there’s a lot here. I’m slightly concerned that either a) the feature is much more complex than anticipated b) we’ve made it more complicated than it needs to be

I think it's easy to conflate word count with complexity here, but really, the feature is quite simple. We have the 'hard' parts solved already-- encoding the events, the topology pipeline, streaming data back to clients via the API, etc.

Much of the discourse is specific to one implementation detail of whether it should live - or be augmented by - topology. The above commentary has been helping for providing a clearer starting path to experiment with, and I think that's good enough for now.

I'm not concerned with the implementation. This is just necessary discourse to whittle down the most efficient of several approaches.

> The use case is production setup and debugging. The ability to inspect events running through a Vector instance to investigate abnormal behavior, usually due to malformed data. That’s it really. This feature is very important for a GUI intended to build production pipelines, but it sounds like we may solve that differently?

The use-case is understood. It wasn't initially clear whether we were talking about observing _raw bytes_ (which would literally allow us to "see" what a sink was seeing) or encoded `Event::Log`s. Some of the earlier requirement-teasing was to do with that. That's been concluded, so you can ignore the first half of the commentary.

The API is very well suited to getting this data out to the client, and the UI work @sghall has done and is doing, is aligned well to consume this data. The hard parts were solved by earlier metrics work. We inherit the type safety, streaming data over WebSockets, etc. This is not new ground we'll need to cover.

What's left to decide is really just what gets the data out of topology and into the API to consume, is a sink, or some kind of external layer. I have some good ideas for that based on the feedback already received, certainly enough to start working toward a prototype.

> Given the above, a light RFC is in order. If we find that the RFC is overly complex, or the discussion is heavy, we should consider not doing this project as it’s value would not offset the cost.

There's nothing particularly burdensome about this feature that I think should warrant shelving it. Certain things just need discussing more than others to whittle 10 possible paths, down to 1-2.

I would suggest an RFC after creating a prototype, if the prototype itself doesn't already align expectations.

I've found that process to be very useful when initially engineering the API. These types of features tend to have a big visual element to them that is easier to explain through demonstration. I think we risk feeling different parts of the elephant, and not being entirely sure what we're each describing otherwise. Easier to see it working in some form first.

**Comment by binarylogic** [comment]
Sounds good! I’m ooo and on my phone so apologies if my response was a little short. I agree with the discourse, I think it’s great and exactly what we should discuss before work. I did get the sense that the amount of discourse was reflective of the complexity,  but if that’s not true then that’s better.

To step back, outside of the nice benefit of being able to inspect events in a _single_ Vector instance, the main value was to lay a foundation for setting up and debugging fleets, similar to top. But if that’s not the case, then the cost threshold comes way down. It’s not the way the feature isn’t valuable, but taking fleet management out of the equation reduces it considerably. Does that make sense?

**Comment by leebenson** [comment]
> To step back, outside of the nice benefit of being able to inspect events in a single Vector instance, the main value was to lay a foundation for setting up and debugging fleets, similar to top. But if that’s not the case, then the cost threshold comes way down. It’s not that the feature isn’t valuable, but taking fleet management out of the equation reduces it considerably. Does that make sense?

Yes, this is still very much the case -- with or without fleet management.

The bulk of this issue is focused on copying events _out_ of the fixed topology pipeline, to external subscribers. This is true for single Vector observability via the API, as much as it is for a fleet of clients. Even if we haven't decided exactly what fleet management looks like, this work is likely to be useable in whatever direction we evolve into, since it's really just a general pub/sub pattern.

The API for the current single instance Vector is its first client.

A lot of this discussion here is cantered around where that integration point should be, and how to make subscribing to those events efficient.

Given that `tap`'s defining characteristic is the ability to query individual component streams, it's not a case of simply dropping in another sink; our sinks lack any meta/awareness of where the event originated. Likewise, it's not a case of just subscribing to _all_ events, since it's likely that tap won't always have active subscribers.

That's much of what this thread has been about (that and integrating remap!)

After some experimentation, I think I have a decent starting point for observing events efficiently, without altering topology or sink definitions, and in a way that works well with our current clients -- `vector tap` as a CLI command, and being able to both query for specific events _and_ filter in the UI via the API.
